File: api/src/main/java/net/asch/bulkit/api/registry/ResourceType.java
package net.asch.bulkit.api.registry;

import net.asch.bulkit.api.data.ResourceIdentifier;
import net.asch.bulkit.api.item.Disk;
import net.minecraft.core.Registry;
import net.minecraft.core.component.DataComponentType;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.neoforged.neoforge.capabilities.BlockCapability;
import net.neoforged.neoforge.capabilities.ICapabilityProvider;
import net.neoforged.neoforge.capabilities.ItemCapability;
import net.neoforged.neoforge.capabilities.RegisterCapabilitiesEvent;
import net.neoforged.neoforge.registries.DeferredHolder;
import net.neoforged.neoforge.registries.DeferredItem;
import net.neoforged.neoforge.registries.DeferredRegister;
import org.jetbrains.annotations.Nullable;

import java.util.function.Function;
import java.util.function.Supplier;

public class ResourceType<T> {
    public final String key;
    public final DeferredHolder<DataComponentType<?>, DataComponentType<ResourceIdentifier<T>>> id;
    public final DeferredItem<Disk> disk;
    private final IDiskCapabilityRegister diskCapRegister;
    private final IDriveNetworkViewCapabilityRegister driveNetworkViewCapRegister;

    private ResourceType(String key, DeferredHolder<DataComponentType<?>, DataComponentType<ResourceIdentifier<T>>> id, DeferredItem<Disk> disk, IDiskCapabilityRegister diskCapRegister, IDriveNetworkViewCapabilityRegister driveNetworkViewCapRegister) {
        this.key = key;
        this.id = id;
        this.disk = disk;
        this.diskCapRegister = diskCapRegister;
        this.driveNetworkViewCapRegister = driveNetworkViewCapRegister;
    }

    public void registerDiskCapability(RegisterCapabilitiesEvent event) {
        diskCapRegister.register(event);
    }

    public <BE extends BlockEntity> void registerDriveNetworkViewCapability(RegisterCapabilitiesEvent event, BlockEntityType<BE> blockEntityType) {
        driveNetworkViewCapRegister.register(event, blockEntityType);
    }

    public static class Builder<T> implements Supplier<ResourceType<T>> {
        final String key;
        private final DeferredRegister.DataComponents dataComponents;
        private final DeferredRegister.Items items;
        private DeferredHolder<DataComponentType<?>, DataComponentType<ResourceIdentifier<T>>> id;
        private DeferredItem<Disk> disk;
        private IDiskCapabilityRegister diskCapRegister;
        private IDriveNetworkViewCapabilityRegister driveNetworkViewCapRegister;

        public Builder(String key, DeferredRegister.DataComponents dataComponents, DeferredRegister.Items items) {
            this.key = key;
            this.dataComponents = dataComponents;
            this.items = items;
        }

        public Builder<T> registry(Registry<T> registry) {
            id = dataComponents.registerComponentType("resource_" + key, (builder) -> builder.persistent(ResourceIdentifier.codec(registry)).networkSynchronized(ResourceIdentifier.streamCodec(registry)).cacheEncoding());
            return this;
        }

        public Builder<T> defaultDisk() {
            return disk(Disk::new);
        }

        public Builder<T> disk(Function<Item.Properties, Disk> sup) {
            disk = items.registerItem("disk_" + key, sup);
            return this;
        }

        public <H> Builder<T> diskHandler(ItemCapability<H, @Nullable Void> cap, ICapabilityProvider<ItemStack, @Nullable Void, H> provider) {
            diskCapRegister = (RegisterCapabilitiesEvent event) -> event.registerItem(cap, provider, disk);
            return this;
        }

        public <H, C> Builder<T> driveNetworkViewHandler(BlockCapability<H, C> cap, ICapabilityProvider<BlockEntity, C, H> provider) {
            driveNetworkViewCapRegister = (RegisterCapabilitiesEvent event, BlockEntityType<? extends BlockEntity> blockEntityType) -> event.registerBlockEntity(cap, blockEntityType, provider);
            return this;
        }

        @Override
        public ResourceType<T> get() {
            return new ResourceType<>(key, id, disk, diskCapRegister, driveNetworkViewCapRegister);
        }
    }

    @FunctionalInterface
    public interface IDiskCapabilityRegister {
        void register(RegisterCapabilitiesEvent event);
    }

    @FunctionalInterface
    public interface IDriveNetworkViewCapabilityRegister {
        void register(RegisterCapabilitiesEvent event, BlockEntityType<? extends BlockEntity> blockEntityType);
    }
}

File: api/src/main/kotlin/net/asch/bulkit/api/registry/ResourceType.kt
package net.asch.bulkit.api.registry

import net.asch.bulkit.api.capability.Capabilities
import net.asch.bulkit.api.data.ResourceIdentifier
import net.asch.bulkit.api.item.Disk
import net.asch.bulkit.api.registry.ResourceType.IDiskCapabilityRegister
import net.asch.bulkit.api.registry.ResourceType.IDriveNetworkViewCapabilityRegister
import net.minecraft.core.Registry
import net.minecraft.core.component.DataComponentType
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.neoforged.neoforge.capabilities.BlockCapability
import net.neoforged.neoforge.capabilities.ICapabilityProvider
import net.neoforged.neoforge.capabilities.ItemCapability
import net.neoforged.neoforge.capabilities.RegisterCapabilitiesEvent
import net.neoforged.neoforge.registries.DeferredHolder
import net.neoforged.neoforge.registries.DeferredItem
import net.neoforged.neoforge.registries.DeferredRegister
import java.util.function.Function
import java.util.function.Supplier

class ResourceType<T>(
    val key: String,
    val id: DeferredHolder<DataComponentType<*>, DataComponentType<ResourceIdentifier<T>>>,
    val disk: DeferredItem<Disk>,
    private val diskCapRegister: IDiskCapabilityRegister,
    private val driveNetworkViewCapRegister: IDriveNetworkViewCapabilityRegister
) {
    fun registerDiskCapability(event: RegisterCapabilitiesEvent) = diskCapRegister.register(event)

    fun <BE : BlockEntity> registerDriveNetworkViewCapability(
        event: RegisterCapabilitiesEvent, blockEntityType: BlockEntityType<BE>
    ) = driveNetworkViewCapRegister.register(event, blockEntityType)

    class Builder<T>(
        val key: String, val dataComponents: DeferredRegister.DataComponents, private val items: DeferredRegister.Items
    ) : Supplier<ResourceType<T>> {
        lateinit var id: DeferredHolder<DataComponentType<*>, DataComponentType<ResourceIdentifier<T>>>
        private lateinit var disk: DeferredItem<Disk>
        private lateinit var diskCapRegister: IDiskCapabilityRegister
        private lateinit var driveNetworkViewCapRegister: IDriveNetworkViewCapabilityRegister

        fun registry(registry: Registry<T>): Builder<T> {
            id = dataComponents.registerComponentType("resource_$key") { builder ->
                builder.persistent(ResourceIdentifier.codec(registry))
                    .networkSynchronized(ResourceIdentifier.streamCodec(registry)).cacheEncoding()
            }

            return this
        }

        fun defaultDisk(): Builder<T> = disk { properties -> Disk(properties) }
        fun disk(sup: Function<Item.Properties, Disk>): Builder<T> {
            disk = items.registerItem("disk_$key", sup)
            return this
        }

        fun <H> diskHandler(
            cap: ItemCapability<H, Void?>,
            resourceProvider: ICapabilityProvider<ItemStack, Void?, H>
        ): Builder<T> {
            diskCapRegister = IDiskCapabilityRegister { event ->
                event.registerItem(cap, resourceProvider, disk)
            }
            return this
        }

        fun <H, C> driveNetworkViewHandler(
            cap: BlockCapability<H, C>, provider: ICapabilityProvider<BlockEntity, C, H>
        ): Builder<T> {
            driveNetworkViewCapRegister = IDriveNetworkViewCapabilityRegister { event, blockEntityType ->
                event.registerBlockEntity(
                    cap, blockEntityType, provider
                )
            }
            return this
        }

        override fun get(): ResourceType<T> {
            return ResourceType(key, id, disk, diskCapRegister, driveNetworkViewCapRegister)
        }
    }


    fun interface IDiskCapabilityRegister {
        fun register(event: RegisterCapabilitiesEvent)
    }

    fun interface IDriveNetworkViewCapabilityRegister {
        fun register(event: RegisterCapabilitiesEvent, blockEntityType: BlockEntityType<out BlockEntity>)
    }
}

fun ResourceType.Builder<Unit>.resourceLess(): ResourceType.Builder<Unit> {
    id =
        dataComponents.registerComponentType("resource_$key") { builder -> builder.persistent(ResourceIdentifier.RESOURCELESS_CODEC) }
    return this
}